### PR TITLE
5983 update to use np.linalg for the small affine inverse

### LINKED ISSUE #5696: multi-thread context torch.inverse
```
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/localhome/Projects/monailabel/monailabel/interfaces/utils/app.py", line 132, in <module>
    run_main()
  File "/localhome/Projects/monailabel/monailabel/interfaces/utils/app.py", line 121, in run_main
    result = a.batch_infer(request)
  File "/localhome/Projects/monailabel/monailabel/interfaces/app.py", line 338, in batch_infer
    return self._batch_infer(request, datastore if datastore else self.datastore(), self.infer)
  File "/localhome/Projects/monailabel/monailabel/interfaces/tasks/batch_infer.py", line 100, in __call__
    res = future.result()
  File "/usr/lib/python3.10/concurrent/futures/_base.py", line 458, in result
    return self.__get_result()
  File "/usr/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
    raise self._exception
  File "/usr/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/localhome/Projects/monailabel/monailabel/interfaces/tasks/batch_infer.py", line 124, in run_infer_task
    r = infer(req, datastore)
  File "/localhome/Projects/monailabel/monailabel/interfaces/app.py", line 300, in infer
    result_file_name, result_json = task(request)
  File "/localhome/Projects/monailabel/monailabel/tasks/infer/basic_infer.py", line 297, in __call__
    data = self.run_pre_transforms(data, pre_transforms)
  File "/localhome/Projects/monailabel/monailabel/tasks/infer/basic_infer.py", line 388, in run_pre_transforms
    return run_transforms(data, transforms, log_prefix="PRE", use_compose=False)
  File "/localhome/Projects/monailabel/monailabel/interfaces/utils/transform.py", line 106, in run_transforms
    data = t(data)
  File "/localhome/Projects/MONAI/monai/transforms/spatial/dictionary.py", line 416, in __call__
    d[key] = self.spacing_transform(
  File "/localhome/Projects/MONAI/monai/utils/deprecate_utils.py", line 221, in _wrapper
    return func(*args, **kwargs)
  File "/localhome/Projects/MONAI/monai/transforms/spatial/array.py", line 606, in __call__
    data_array = self.sp_resample(
  File "/localhome/Projects/MONAI/monai/utils/deprecate_utils.py", line 221, in _wrapper
    return func(*args, **kwargs)
  File "/localhome/Projects/MONAI/monai/transforms/spatial/array.py", line 315, in __call__
    img = affine_xform(img.unsqueeze(0), theta=xform, spatial_size=spatial_size).squeeze(0)
  File "/localhome/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/localhome/Projects/MONAI/monai/networks/layers/spatial_transforms.py", line 556, in forward
    theta = to_norm_affine(
  File "/localhome/Projects/MONAI/monai/networks/utils.py", line 279, in to_norm_affine
    return src_xform @ affine @ torch.inverse(dst_xform)
RuntimeError: lazy wrapper should be called at most once
```

### LINKED ISSUE #5983: Torch.inverse should be replaced to achieve stable inverse result.
**Describe the bug**

`Spacing` transform relies on the `torch.inverse()` to compute the affine matrix. However, the operation has errors compared to `numpy`. As a result, the integration test `test_integration_segmentation3d` is affected by the changes between PyTorch versions.

As an example, `torch.inverse()` provides different results between PyTorch 22.09 and 22.11. Steps to reproduce are listed below. The error seems to be at the range of 1e-7 on ampere GPUs.
Furthermore, we use `numpy` function `np.linalg.inv()` as the gold standard.
For the particular input, we see the error in 22.11/22.12/23.01 is 3 times higher than 22.09.
We tried to turn off TF32, set the matmul precision to highest (FP32) in PyTorch containers (22.11-23.01), and the result is the same.

Result in 22.09:
 
Input tensor: 
tensor([[[1.8692e-02, 0.0000e+00, 0.0000e+00, -9.9065e-01],
         [0.0000e+00, 1.2500e-02, 0.0000e+00, -9.9375e-01],
         [0.0000e+00, 0.0000e+00, 1.0989e-02, -9.9451e-01],
         [0.0000e+00, 0.0000e+00, 0.0000e+00, 1.0000e+00]]])
Output of torch.inverse in PyTorch 22.09: 
tensor([[[5.3500e+01, 0.0000e+00, -0.0000e+00, 5.3000e+01],
         [0.0000e+00, 8.0000e+01, -0.0000e+00, 7.9500e+01],
         [0.0000e+00, 0.0000e+00, 9.1000e+01, 9.0500e+01],
         [0.0000e+00, 0.0000e+00, 0.0000e+00, 1.0000e+00]]])
Error compared with numpy
tensor(1.1921e-07)

Result in 22.11
 
Input tensor: 
tensor([[[1.8692e-02, 0.0000e+00, 0.0000e+00, -9.9065e-01],
         [0.0000e+00, 1.2500e-02, 0.0000e+00, -9.9375e-01],
         [0.0000e+00, 0.0000e+00, 1.0989e-02, -9.9451e-01],
         [0.0000e+00, 0.0000e+00, 0.0000e+00, 1.0000e+00]]])
Output of torch.inverse in PyTorch 22.11: 
tensor([[[ 5.3500e+01, -2.2252e-06, -1.3113e-06,  5.3000e+01],
         [ 6.3144e-06,  8.0000e+01, -1.9670e-06,  7.9500e+01],
         [-6.6143e-06, -9.4399e-06,  9.1000e+01,  9.0500e+01],
         [-7.4506e-09, -4.1986e-08, -2.4742e-08,  1.0000e+00]]])
Error compared with numpy
tensor(3.4737e-07)

**To Reproduce**
1. Run the snippet in shell script to get result in PyTorch 22.09 container.
 ```
docker run -it --rm --gpus=all -e NVIDIA_TF32_OVERRIDE=0 nvcr.io/nvidia/pytorch:22.09-py3 bash -c "python << EOF
import torch
import numpy as np
torch.set_printoptions(sci_mode=True)
torch.set_float32_matmul_precision(\"highest\")
C = torch.tensor([[[ 0.01869158819317817687988281250000,
           0.00000000000000000000000000000000,
           0.00000000000000000000000000000000,
          -0.99065423011779785156250000000000],
         [ 0.00000000000000000000000000000000,
           0.01250000018626451492309570312500,
           0.00000000000000000000000000000000,
          -0.99374997615814208984375000000000],
         [ 0.00000000000000000000000000000000,
           0.00000000000000000000000000000000,
           0.01098901126533746719360351562500,
          -0.99450546503067016601562500000000],
         [ 0.00000000000000000000000000000000,
           0.00000000000000000000000000000000,
           0.00000000000000000000000000000000,
           1.00000000000000000000000000000000]]])

print(\"Input tensor: \")
print(C)
print(\"Output of torch.inverse in PyTorch 22.09: \")
print(torch.inverse(C))
print(\"Error compared with numpy\")
print((torch.tensor(np.array(C) @ np.linalg.inv(np.array(C))) - C @ torch.inverse(C)).abs().sum())
EOF
"
```

2. Run the same snippet in a different container later than 22.11. Here we use 22.11 for example:
```
docker run -it --rm --gpus=all -e NVIDIA_TF32_OVERRIDE=0 nvcr.io/nvidia/pytorch:22.11-py3 bash -c "python << EOF
import torch
import numpy as np
torch.set_printoptions(sci_mode=True)
torch.set_float32_matmul_precision(\"highest\")
C = torch.tensor([[[ 0.01869158819317817687988281250000,
           0.00000000000000000000000000000000,
           0.00000000000000000000000000000000,
          -0.99065423011779785156250000000000],
         [ 0.00000000000000000000000000000000,
           0.01250000018626451492309570312500,
           0.00000000000000000000000000000000,
          -0.99374997615814208984375000000000],
         [ 0.00000000000000000000000000000000,
           0.00000000000000000000000000000000,
           0.01098901126533746719360351562500,
          -0.99450546503067016601562500000000],
         [ 0.00000000000000000000000000000000,
           0.00000000000000000000000000000000,
           0.00000000000000000000000000000000,
           1.00000000000000000000000000000000]]])

print(\"Input tensor: \")
print(C)
print(\"Output of torch.inverse in PyTorch 22.11: \")
print(torch.inverse(C))
print(\"Error compared with numpy\")
print((torch.tensor(np.array(C) @ np.linalg.inv(np.array(C))) - C @ torch.inverse(C)).abs().sum())
EOF
"
```
**Expected behavior**

**Screenshots**

**Environment**

Ensuring you use the relevant python executable, please paste the output of:

```
python -c 'import monai; monai.config.print_debug_info()'
```

**Additional context**

### PR DESCRIPTION
**Pull request by wyli**
Signed-off-by: Wenqi Li <wenqil@nvidia.com>

Fixes #5983 
Fixes #5696 

### Description
- torch.inverse was less stable across different base image versions
- there's a bug for loading the module https://github.com/pytorch/pytorch/issues/90613
this therefore also closes https://github.com/Project-MONAI/MONAI/issues/5696

this PR tries to move away from the torch.linalg APIs for small inverse tasks where there's no need for making them differentiable 
Screenshot 2023-02-09 at 23 42 15

Screenshot 2023-02-13 at 14 47 50

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/networks/utils.py [hunk 1]
```diff
@@ -264,8 +265,8 @@ def to_norm_affine(
         raise ValueError(f"affine suggests {sr}D, got src={len(src_size)}D, dst={len(dst_size)}D.")
 
     src_xform = normalize_transform(src_size, affine.device, affine.dtype, align_corners, zero_centered)
-    dst_xform = normalize_transform(dst_size, affine.device, affine.dtype, align_corners, zero_centered)
-    return src_xform @ affine @ torch.inverse(dst_xform)
+    dst_xform = normalize_transform(dst_size, "cpu", affine.dtype, align_corners, zero_centered)
```
Comment (Nic-Ma) [inline monai/networks/utils.py:268]: Is there any case in MONAI that the `affine` is on GPU? It that case, this `GPU -> CPU -> GPU` will be extremely slow and may become a bottleneck.

Thanks.

File: monai/networks/utils.py [hunk 2]
```diff
@@ -264,8 +265,8 @@ def to_norm_affine(
...
+    return src_xform @ affine @ convert_to_dst_type(np.linalg.inv(dst_xform.numpy()), dst=affine)[0]
```
Comment (Nic-Ma) [inline monai/networks/utils.py:269]: Could you please add a comment here to explain why we convert to numpy for future understanding.

Thanks.

File: monai/transforms/spatial/array.py [hunk 3]
```diff
@@ -293,12 +290,12 @@ def __call__(
...
+            dst_xform_1 = normalize_transform(spatial_size, "cpu", xform.dtype, True, True)[0].numpy()  # to (-1, 1)
             if not align_corners:
-                norm = create_scale(spatial_rank, [(max(d, 2) - 1) / d for d in spatial_size], xform.device, "torch")
-                dst_xform_1 = norm.to(xform.dtype) @ dst_xform_1  # type: ignore  # scaling (num_step - 1) / num_step
-            dst_xform_d = normalize_transform(spatial_size, xform.device, xform.dtype, align_corners, False)[0]
-            xform = xform @ torch.inverse(dst_xform_d) @ dst_xform_1
+                norm = create_scale(spatial_rank, [(max(d, 2) - 1) / d for d in spatial_size])
+                dst_xform_1 = norm.astype(float) @ dst_xform_1  # type: ignore  # scaling (num_step - 1) / num_step
+            dst_xform_d = normalize_transform(spatial_size, "cpu", xform.dtype, align_corners, False)[0].numpy()
+            xform @= convert_to_dst_type(np.linalg.solve(dst_xform_d, dst_xform_1), xform)[0]
```
Comment (Nic-Ma) [inline monai/transforms/spatial/array.py:298]: Same comment as the above.

Thanks.

File: tests/testing_data/integration_answers.py [hunk 4]
```diff
@@ -493,6 +493,73 @@
...
+                0.5373653262853623,
+                0.46776085495948794,
+                0.4422474503517151,
+                0.43667820692062376,
+                0.42639826238155365,
+                0.4158218264579773,
+            ]
+        }
+    },
+    {  # test answers for PyTorch 21.10 mixed readers
```
Comment (Nic-Ma) [inline tests/testing_data/integration_answers.py:509]: May I know what "mixed readers" mean here?

Thanks.

**Comment by mingxin-zheng** [comment]
I found some issue when I test this fix on PyTorch 22.10 on A100. The integration test `test_integration_segmentation_3d.py` did not pass.

[2209.txt](https://github.com/Project-MONAI/MONAI/files/10704615/2209.txt)
[2210.txt](https://github.com/Project-MONAI/MONAI/files/10704616/2210.txt)
[2211.txt](https://github.com/Project-MONAI/MONAI/files/10704617/2211.txt)
[2212.txt](https://github.com/Project-MONAI/MONAI/files/10704618/2212.txt)

File: monai/networks/utils.py (hunk 2 above)
Comment (wyli) [reply monai/networks/utils.py:269]: I think @mingxin-zheng can create a ticket on github for the technical details, with similar content provided on nvbugs?

File: tests/testing_data/integration_answers.py (hunk 4 above)
Comment (wyli) [reply tests/testing_data/integration_answers.py:509]: yes that means in this for loop of testing different combinations, not all of them failed, this answer matches some of them...

https://github.com/Project-MONAI/MONAI/blob/5657b8f2d297e2d2b09017cf9f57dad2a72186f4/tests/test_integration_segmentation_3d.py#L291-L292

File: monai/networks/utils.py (hunk 1 above)
Comment (wyli) [reply monai/networks/utils.py:268]: this variable is constructed on cpu, and then moved to the same device as `affine`, so I think the overhead is reasonably small...

File: monai/networks/utils.py (hunk 2 above)
Comment (mingxin-zheng) [reply monai/networks/utils.py:269]: Okay! Here it is:
https://github.com/Project-MONAI/MONAI/issues/5983

Comment (wyli) [reply monai/networks/utils.py:269]: Thanks, I've linked the pr to the ticket.

**Review by mingxin-zheng — APPROVED** [review]

File: monai/networks/utils.py (hunk 2 above)
Comment (Nic-Ma) [reply monai/networks/utils.py:269]: Could you please just add 1 line comment to help others understand why we convert it to numpy array here?

Thanks.

Comment (wyli) [reply monai/networks/utils.py:269]: let me put the information on the conntected ticket.

**Review by Nic-Ma — APPROVED** [review]

**Pull request merged**